sdk/samples/iot/aziot_realtek_amebaD/aziot_realtek_amebaD.ino

```diff
@@ -0,0 +1,379 @@
+static const char* host = IOT_CONFIG_IOTHUB_FQDN;
+static const char* device_id = IOT_CONFIG_DEVICE_ID;
+static const az_span device_key = AZ_SPAN_FROM_STR(IOT_CONFIG_DEVICE_KEY);
+static const int mqtt_port = 8883;
+
+static WiFiUDP ntp_udp_client;
+static WiFiSSLClient wifi_client;
+static PubSubClient mqtt_client(wifi_client);
+static az_iot_hub_client hub_client;
+static char sas_token[200];
+static size_t sas_token_length;
+
+static uint8_t signature[512];
+static unsigned long next_telemetry_send_time_ms = 0;
+static char telemetry_topic[128];
+static uint8_t telemetry_payload[100];
+static uint32_t telemetry_send_count = 0;
+static unsigned char* ca_pem_nullterm;
+
+extern "C"{
+  extern int mbedtls_base64_decode( unsigned char *dst, size_t dlen, size_t *olen, const unsigned char *src, size_t slen);
+  extern int rom_hmac_sha256(const u8 *key, size_t key_len, const u8 *data, size_t data_len, u8 *mac);
+  extern int mbedtls_base64_encode( unsigned char *dst, size_t dlen, size_t *olen, const unsigned char *src, size_t slen);
+}
+static void createNullTerminatedRootCert()
+{
+  ca_pem_nullterm = (unsigned char*)malloc(ca_pem_len + 1);
+
+  if (ca_pem_nullterm == NULL)
+  {
+    Serial.println("Failed allocating memory for null-terminated root ca");
+  }
+  else {}
+  
+  memcpy(ca_pem_nullterm, ca_pem, ca_pem_len);
+  ca_pem_nullterm[ca_pem_len] = '\0';
+}
+
+
+static void connectToWiFi()
+{
+  Serial.begin(115200);
+  Serial.println();
+
+  Serial.print("Connecting to WIFI SSID ");
+  Serial.println(ssid);
+
+  WiFi.begin((char*)ssid, (char*)password);
+  while (WiFi.status() != WL_CONNECTED)
+  {
+    delay(500);
+    Serial.print(".");
+  }
+
+  Serial.println("");
+  Serial.print("WiFi connected, IP address: ");
+  Serial.println(WiFi.localIP());
+}
+
+void receivedCallback(char* topic, byte* payload, unsigned int length)
+{
+  Serial.print("Received [");
+  Serial.print(topic);
+  Serial.print("]: ");
+  for (unsigned int i = 0; i < length; i++)
+  {
+    Serial.print((char)payload[i]);
+  }
+  Serial.println("");
+}
+
+static void initializeClients()
+{
+  Serial.println("Initializing MQTT client");
+  
+  wifi_client.setRootCA(ca_pem_nullterm);
+
+  if (az_result_failed(az_iot_hub_client_init(
+          &hub_client,
+          az_span_create((uint8_t*)host, strlen(host)),
+          az_span_create((uint8_t*)device_id, strlen(device_id)),
+          NULL)))
+  {
+    Serial.println("Failed initializing Azure IoT Hub client");
+    return;
+  }
+
+  mqtt_client.setServer(host, mqtt_port);
+  mqtt_client.setCallback(receivedCallback);
+
+  Serial.println("MQTT client initialized");
+}
+
+int64_t iot_sample_get_epoch_expiration_time_from_minutes(uint32_t minutes)
+{
+
+  long ts = 0;
+  long tus = 0;
+  unsigned int ttk = 0;
+
+  //it should be ok to do init more than one time. It'll handle inside sntp_init().
+  sntp_init();
+
+  sntp_get_lasttime(&ts, &tus, &ttk);
+  
+  while(ts == 0){
+    vTaskDelay(1000 / portTICK_RATE_MS);
```

Comment:
>1000 / portTICK_RATE_MS [](start = 15, length = 23)

is there no function called ms_to_ticks?

Reply:
This code has been provided directly by the Realtek team. I will keep it as is in case we need to update this later or use their functions directly from their library.